WASM Example

Hi.

Qt [now supports compiling to WASM](https://doc.qt.io/qt-5/wasm.html). Are we able to do this with CXX-Qt?
If this is true. I think we need to add a WASM example.

Thanks for CXX-Qt.

Based on the qml_minimal example, built with Qt 6.4.2, emscripten 3.1.14 and rust 1.67.0 with `wasm32-unknown-emscripten`:
![image](https://user-images.githubusercontent.com/1146309/218358873-9071627e-03b8-4f89-aea5-bd219874b33b.png)
Success, _but_:

I had to use `target_link_options(${APP_NAME}_lib INTERFACE "SHELL:-Wl,--whole-archive $<TARGET_FILE:${CRATE}-static> -Wl,--no-whole-archive")` instead of `"$<LINK_LIBRARY:WHOLE_ARCHIVE,${CRATE}-static>"`, otherwise CMake complained that
> Feature 'WHOLE_ARCHIVE', specified through generator-expression
>  '$<LINK_LIBRARY>' to link target 'example_qml_minimal', is not supported
>  for the 'CXX' link language.

Remove `QmlImportScanner` from Qt components since it's not present there in cross-builds (there is probably a way to include the x86 one but it's almost 4am :)

And (the bad one) I had to comment out the `assert_alignment_and_size` macro to get `cxx-qt-lib` to compile, otherwise I get
```
cargo:warning=src/types/qcolor.cpp:17:1: error: static_assert failed due to requirement 'sizeof(QColor) == (sizeof(unsigned long[2]))' "unexpected QColor size!"
  cargo:warning=assert_alignment_and_size(QColor, alignof(std::size_t), sizeof(std::size_t[2]));
  cargo:warning=^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=src/types/assertion_utils.h:13:3: note: expanded from macro 'assert_alignment_and_size'
  cargo:warning=  static_assert(sizeof(TYPE) == (SIZE), "unexpected " #TYPE " size!");
  cargo:warning=  ^             ~~~~~~~~~~~~~~~~~~~~~~
```
```
  cargo:warning=src/types/qvariant.cpp:27:1: error: static_assert failed due to requirement 'alignof(QVariant) <= (alignof(unsigned long))' "unexpectedly large QVariant alignment!"
  cargo:warning=assert_alignment_and_size(QVariant,
  cargo:warning=^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=src/types/assertion_utils.h:11:3: note: expanded from macro 'assert_alignment_and_size'
  cargo:warning=  static_assert(alignof(TYPE) <= (ALIGNMENT),                                  \
  cargo:warning=  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=src/types/qvariant.cpp:27:1: error: static_assert failed due to requirement 'sizeof(QVariant) == (sizeof(unsigned long[4]))' "unexpected QVariant size!"
  cargo:warning=assert_alignment_and_size(QVariant,
  cargo:warning=^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=src/types/assertion_utils.h:13:3: note: expanded from macro 'assert_alignment_and_size'
  cargo:warning=  static_assert(sizeof(TYPE) == (SIZE), "unexpected " #TYPE " size!");
  cargo:warning=  ^             ~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=2 errors generated.
```

I'll investigate why the assertions fail next, I guess something to do with wasm being 32bit only?

This is what I did to the example (keep in mind I went the easy way and used the CMake one): https://github.com/maciek134/cxx-qt/commit/d8780440442122a891f3ce46e85ae7e1fbb6ca81
Updating `cxx-qt` to c48dca7 fixed the assertions too, so no ugly hacks anymore :)

I used emscripten 3.1.4, Qt 6.4.2 (Qt downloaded from their website, not recompiled). Built with
```
$ ~/Qt/6.4.2/wasm_32/bin/qt-cmake -B build .
$ cmake --build ./build
```

Something is still wrong with the order somewhere, since on first compilation I get
```
/home/klh/Workspace/qt6wasm/cpp/main.cpp:14:10: fatal error: 'cxx-qt-gen/my_object.cxxqt.h' file not found                                                                                                                                                                                
#include "cxx-qt-gen/my_object.cxxqt.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
And I guess it will always be one compilation old after that.

--- a/.github/workflows/github-cxx-qt-tests.yml
+++ b/.github/workflows/github-cxx-qt-tests.yml
@@ -88,6 +88,135 @@ jobs:
           token: ${{ secrets.CODECOV_TOKEN }}
           verbose: true
 
+  build-wasm:
+    name: Ubuntu 22.04 (wasm_32) Qt 6
+    needs: [clang_format, license_check, rust_format_check, markdown_lint, shellcheck]
+    runs-on: ubuntu-22.04
+    env:
+      SCCACHE_CACHE_SIZE: 600M
+    steps:
+      - name: "Checkout repository"
+        uses: actions/checkout@v4
+      - name: Setup toolchain
+        run: |
+          rustup component add clippy rustfmt
+          rustup target add wasm32-unknown-emscripten
+      - name: "Rust tools cache"
+        uses: actions/cache@v4
+        id: rust-tools-cache
+        with:
+          path: |
+            ~/.cargo/bin/sccache
+            ~/.cargo/bin/mdbook
+            ~/.cargo/bin/mdbook-linkcheck
+          key: ubuntu-22.04_sccache-0.7.6_mdbook-0.4.36_mdbook-linkcheck-0.7.7
+      - name: "Build Rust tools"
+        if: steps.rust-tools-cache.outputs.cache-hit != 'true'
+        # Do not build with storage backends enabled, we only need local
+        run: |
+          cargo install --no-default-features sccache
+          cargo install mdbook mdbook-linkcheck
+      # We want our compiler cache to always update to the newest state.
+      # The best way for us to achieve this is to **always** update the cache after every landed commit.
+      # That way it will closely follow our development.
+      # And if a PR diverges a lot with its cache that's not a big deal, as it will be merged eventually.
+      #
+      # This is a workaround for the fact that GH doesn't support updating existing caches.
+      # See: https://github.com/azu/github-actions-overwrite-cache-example
+      #
+      # Ideally we'd like to use this:
+      # - name: "Compiler cache"
+      #   uses: actions/cache@v4
+      #   with:
+      #     update: true <------- THIS DOESN'T EXIST YET
+      #     path: /home/runner/.cache/sccache
+      #     key: "Ubuntu 22.04 (wasm_32) Qt6_compiler_cache"
+      - name: "Restore Compiler Cache"
+        id: compiler-cache-restore
+        uses: actions/cache/restore@v4
+        with:
+          path: /home/runner/.cache/sccache
+          key: "Ubuntu 22.04 (wasm_32) Qt6_compiler_cache"
+      - name: "emsdk cache"
+        uses: actions/cache@v4
+        id: emsdk-cache
+        with:
+          path: ./emsdk
+          key: emsdk_2.0.14
+      - name: "Setup emsdk"
+        if: steps.emsdk-cache.outputs.cache-hit != 'true'
+        run: |
+          git clone https://github.com/emscripten-core/emsdk
+          ./emsdk/emsdk install 2.0.14
+          ./emsdk/emsdk activate 2.0.14
+      - name: "Install Dependencies"
+        run: sudo apt-get update && sudo apt-get install -y ninja-build clang-format-12
+      - name: Install Qt GCC
+        uses: jurplel/install-qt-action@v3
+        with:
+          aqtversion: '==3.1.*'
+          version: '6.2.4'
+          host: 'linux'
+          target: 'desktop'
+          arch: 'gcc_64'
+          cache: true
+      - name: Install Qt WASM
+        uses: jurplel/install-qt-action@v3
+        with:
+          aqtversion: '==3.1.*'
+          version: '6.2.4'
+          host: 'linux'
+          target: 'desktop'
+          arch: 'wasm_32'
+          tools: 'tools_cmake'
+          cache: true
+      - name: "Configure"
+        # FIXME: ninja is not being found?
+        env:
+          RUSTC_WRAPPER: sccache
+        run: |
+          source ./emsdk/emsdk_env.sh
+          /home/runner/work/cxx-qt/Qt/6.2.4/wasm_32/bin/qt-cmake -DQT_HOST_PATH=/home/runner/work/cxx-qt/Qt/6.2.4/gcc_64 -DBUILD_WASM=ON -Bbuild .
+      - name: "Build"
+        env:
+          RUSTC_WRAPPER: sccache
+        run: cmake --build build --parallel 4
+      - name: Test output files exist
+        run: |
+          test -f ./build/examples/qml_minimal/example_qml_minimal.html
+          test -f ./build/examples/qml_minimal/example_qml_minimal.js
+          test -f ./build/examples/qml_minimal/libqml_minimal.a
+          test -x ./build/examples/qml_minimal/example_qml_minimal.wasm
+          test -f ./build/examples/qml_minimal/qtloader.js
+          test -f ./build/examples/qml_features/example_qml_features.html
+          test -f ./build/examples/qml_features/example_qml_features.js
+          test -f ./build/examples/qml_features/libqml_features.a
+          test -x ./build/examples/qml_features/example_qml_features.wasm
+          test -f ./build/examples/qml_features/qtloader.js
+      - name: "Print compiler cache statistics"
+        run: sccache --show-stats
+      # This is a workaround for the fact that GH doesn't support updating existing caches.
+      # See: https://github.com/azu/github-actions-overwrite-cache-example
+      - name: "Delete previous compiler cache"
+        # Updating th cache doesn't work from forks
+        # So update it once it's merged into the repo
+        if: ${{ steps.compiler-cache-restore.outputs.cache-hit &&  github.event_name == 'push' }}
+        working-directory: /home/runner/cxx-qt
+        continue-on-error: true
+        run: |
+          gh extension install actions/gh-actions-cache
+          gh actions-cache delete "Ubuntu 22.04 (wasm_32) Qt6_compiler_cache" --confirm
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - name: "Save Compiler Cache"
+        # Updating th cache doesn't work from forks
+        # So update it once it's merged into the repo
+        if: ${{ github.event_name == 'push' }}
+        uses: actions/cache/save@v4
+        with:
+          path: /home/runner/.cache/sccache
+          key: "Ubuntu 22.04 (wasm_32) Qt6_compiler_cache"
+
   build:
     # Run after pre checks
     needs: [clang_format, license_check, rust_format_check, markdown_lint, shellcheck]

--- a/book/src/getting-started/6-wasm-builds.md
+++ b/book/src/getting-started/6-wasm-builds.md
@@ -77,7 +77,7 @@ To build for WebAssembly in a project that uses CXX-Qt crates, first follow the 
 
 ### CMakeLists.txt
 
-When compiling a CXX-Qt project for wasm, the Rust target must be set to `wasm32-unknown-emscripten`, and the project must be configured to use POSIX threads. Make sure you have the Emscripten target for `rustc` with `rustup target add wasm-unknown-emscripten`.
+When compiling a CXX-Qt project for wasm, the Rust target must be set to `wasm32-unknown-emscripten`, and the project must be configured to use POSIX threads. Make sure you have the Emscripten target for `rustc` with `rustup target add wasm32-unknown-emscripten`.
 
 ```cmake
 set(Rust_CARGO_TARGET wasm32-unknown-emscripten)

